datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -772,22 +773,22 @@ fn slice_and_maybe_filter(
     filter_opt: Option<&Arc<dyn Array>>,
     offsets: &[usize],
 ) -> Result<Vec<ArrayRef>> {
-    let sliced_arrays: Vec<ArrayRef> = aggr_array
-        .iter()
-        .map(|array| array.slice(offsets[0], offsets[1] - offsets[0]))
-        .collect();
+    let null_array = Arc::new(NullArray::new(0)) as ArrayRef;
+    let mut sliced_arrays: Vec<ArrayRef> = vec![null_array; aggr_array.len()];
```

Comment:
I don't see why this should be faster? 🤔

Reply:
The key point is to get rid of extra allocations https://github.com/apache/arrow-datafusion/issues/5969#issuecomment-1504585917

Reply:
Another optimization that we traverse input collection only once, instead of 2 times in original implementation

Reply:
See https://github.com/apache/arrow-datafusion/issues/5969#issuecomment-1513959517 -- reported performance improvement